### PR TITLE
Add support for the Nix package manager

### LINKED ISSUE #120: [FEATURE] Support nix as a package manager
Packages don't show the number of packages installed by nix.

### PR DESCRIPTION
**Pull request by coolGi69**
Similar to https://github.com/Macchina-CLI/libmacchina/pull/150 except updated to the latest version of libmacchina. (closes https://github.com/Macchina-CLI/libmacchina/issues/120)
Unlike that commit which attempts to estimate the total amount of nix pkgs by counting the files in `/nix/store`, this uses nix's sqlite database (`/nix/var/nix/db/db.sqlite`).\
The sqlite library has also been updated (0.31.1 -> 0.36.0) due to it having the ability to read the path as a uri (so that I can tell sqlite that the database is immutable).

This queries the database using `SELECT COUNT(path) FROM ValidPaths WHERE ultimate=1` (which is equivalent to `nix path-info --all --sigs | grep ultimate | wc -l`) which gives a time of around 10ms, while the equivalent nix command gives around 40ms.

This lookup is grabbing a slightly different metric which gives 739 on my system (nix store giving 1306). So I think nix store is including extra things that don't count as packages (as the `/nix/store` folder stores everything from fetches, sources, man pages, binaries, etc).

(all packages built with `cargo build --release`, and tested with `hyperfine -w 4 -m 500 -N`)
| Type | Speed |
| --- | --- |
| Original Macchina | 5.9 ms ± 0.2 ms |
| SQLite Lookup | 14.2 ms ± 1.7 ms |

### CODE REVIEW

**Comment by Gobidev** [comment]
This looks very interesting. From my experience, invoking `nix-store` was significantly slower than every other readout of `libmacchina`. I am currently using [pretty much the same method as in this PR in pfetch-rs](https://github.com/Gobidev/pfetch-rs/blob/fed3a6d95d9f5e839bf622a53256b7e05b491450/src/lib.rs#L135-L151), but added an extra option to skip it because it is comparatively very slow. Maybe a recent Nix update made this method faster, I will do some testing myself later.

Just out of curiosity, what is your `macchina` runtime if you skip Nix package count?

**Comment by Gobidev** [comment]
I have done a bit of testing in a NixOS Live ISO environment with about 1100 packages installed, and this method is still very slow, with the package count taking about `200ms`.  `pfetch-rs` currently has a total runtime below `10ms` in the same environment when skipping the package count, so this is still a big hit to the runtime. This also only gets worse with more packages installed and on older hardware.

So as much as I would like Nix support to be added, this method just is not fast enough for `libmacchina` in my opinion.

If you are interested, you could try to get the package count in other ways, as I tried in #150, but this probably requires deeper knowledge of the nix source code.

I also had a look [at what `fastfetch` does](https://github.com/fastfetch-cli/fastfetch/blob/999f52b0b7810b027669174de78afae8e6efb61b/src/detection/packages/packages_linux.c#L196), they seem to use the command invocation and cache the result for future queries, which is something we could also think about for `libmacchina`.

**Comment by coolGi2007** [comment]
I think that command is the standard for getting the amount of packages. It's also whats used in [fastfetch](https://github.com/fastfetch-cli/fastfetch/blob/9104db0d2148369e6dea25539b896f068718bf3c/src/detection/packages/packages_linux.c#L232) (which is which is what I compared it to), as well as most stack overflow answers.

<br>

(all packages built with `cargo build --release`, and tested with `hyperfine -w 4 -m 500 -N`)

| Default | | |
| --- | --- | --- |
| Time (mean ± σ) | 2.2 ms ±   0.4 ms | [User: 0.7 ms, System: 1.3 ms] |
| Range (min … max) | 1.7 ms …   3.2 ms | 1005 runs |

| This pull | | |
| --- | --- | --- |
| Time (mean ± σ) | 116.5 ms ±  22.1 ms | [User: 46.6 ms, System: 24.3 ms] |
| Range (min … max) | 97.3 ms … 162.3 ms | 500 runs |

| This pull, but removing nix-user (as pr 150 only gives one value) | | |
| --- | --- | --- |
| Time (mean ± σ) | 50.3 ms ±   9.6 ms | [User: 20.5 ms, System: 11.5 ms] |
| Range (min … max) | 41.2 ms …  69.2 ms | 500 runs |

| This pull, but replacing the method with the one in pr 150 | | |
| --- | --- | --- |
| Time (mean ± σ) | 28.9 ms ±   4.8 ms | [User: 5.0 ms, System: 23.3 ms] |
| Range (min … max) | 25.2 ms …  41.6 ms | 500 runs |

I think my previous test were wrong (I'm not even sure how I got the original results). These numbers seem to be more realistic.

<br>
<br>

In reply to your latest comment, as well as the more realistic results, it is probably best to cache the result somewhere. I'll look into attempting the results a bit later tonight if that's the best way to go.

**Comment by grtcdr** [comment]
Hi, thanks for the contribution. I personally want to see more progress in #150 due to the performance penalty of dishing out to a command-line program.

To respond to @Gobidev on the caching proposal, I believe it shouldn't be libmacchina's responsibility to cache things. I'd rather keep the library as stateless as possible (despite the obvious oxymoron), downstream programs should implement this if they so wish.

**Comment by coolGi2007** [comment]
Quick update:\
Nix stores all its data for stuff in `/nix/var/nix/db/db.sqlite` (defined in the [libstore/local-store](https://github.com/NixOS/nix/blob/2f5fdab06c57c03b9eb5f03bd11e8707026c1676/src/libstore/local-store.cc#L549))\
In it, it contains several tables, but the one we are interested in is the `DerivationOutputs` table, and its `id` column. The quickest way is to just count the amount of items with the id of `dist` giving an extremely close approximation of the value gotten by the original `nix-store` command (`nix-store` giving 1306, and the amount of `dist`s 1033)\
I think I'll just need to look into the code a bit more to see where the extra 300 packages come from so it can be more accurate.

Hopefully this method works and can give good speeds! (well, as good as you'll get with an sqlite database)

<br>

Also, out of curiosity I checked why the nix store command was so slow, and it was due to it doing sorting (which makes sense), as well as several places where they are nesting for loops.

**Comment by Gobidev** [comment]
Nice, that sounds really promising, thank you for putting in the work! I will close #150 in favor of this PR.

**Comment by coolGi2007** [comment]
I think the best sqlite query is `SELECT COUNT(path) FROM ValidPaths WHERE ultimate=1` (which is equivalent to `nix path-info --all --sigs | fgrep ultimate | wc -l`) which gives a time of around 10ms, while that nix command gives around 40ms.

Here are the results of hyperfine:
| Type | Speed |
| --- | --- |
| Original Macchina | 5.9 ms ± 0.2 ms |
| SQLite Lookup | 14.2 ms ± 1.7 ms |

This lookup is grabbing a slightly different metric which gives 739 on my system (nix store giving 1306). So I think nix store is including extra things that don't count as packages (as the `/nix/store` folder stores everything from fetches, sources, man pages, binaries, etc).

<br>

Also, I've updated sqlite from 0.31.1 to 0.36.0.

File: src/linux/mod.rs [hunk 1]
```diff
@@ -942,8 +942,8 @@ impl LinuxPackageReadout {
     /// Returns the number of installed packages for systems
     /// that utilize `nix` as their package manager.
     fn count_nix() -> Option<usize> {
-        // Return the number of installed packages using sqlite (~10ms)
-        // as directly calling nix is a bit more expensive (~40ms)
+        // Return the number of installed packages using nix's database (~10ms)
+        // as directly calling to nix is a bit more expensive (~40ms)
```
Comment (grtcdr) [inline src/linux/mod.rs:946]: The benchmarks you listed in the comments should suffice, we shouldn't really have these in the comments.

**Comment by grtcdr** [comment]
Big thanks for the contribution (TIL about [breaking out of loops with labels](https://doc.rust-lang.org/rust-by-example/flow_control/loop/nested.html), thanks for that too!).

**Comment by Gobidev** [comment]
Thanks again for your work. I didn't have much time to test this, yet, but at least in a NixOS live CD environment, it doesn't really seem to work right now. Your method returns `3` installed packages, while `nix-store -q --requisites /run/current-system/sw | wc -l` returns `1098` and `nix-store -q --requisites ~/.nix-profile | wc -l` returns `15`.

> This lookup is grabbing a slightly different metric which gives 739 on my system (nix store giving 1306). So I think nix store is including extra things that don't count as packages (as the /nix/store folder stores everything from fetches, sources, man pages, binaries, etc).

If the lower count is more accurate than the one returned by `nix-store`, this is of course fine, though I expect that there will be issues reported because of inconsistent package counts across different fetch tools.

I will check your method on a system with NixOS actually installed and on a different distro with Nix installed to see if that gives different results, maybe there is just something weird about the NixOS live CD.

**Comment by coolGi2007** [comment]
It only lists the main installed packages. I've uploaded a video of me adding cargo (to a Mint system that I freshly installed Nix onto), and despite it having several dependencies, it only recognizes it as a single one.\
The result is the same on NixOS, but I didn't have my system with it installed on hand to record.

https://github.com/Macchina-CLI/libmacchina/assets/57488297/84278267-71f7-4959-bcd2-c88cf803728b

(I haven't tested it on the NixOS live env, but my guess is that it only shows 3 as it is registering basically everything under a single meta package)

**Comment by coolGi2007** [comment]
The other option for an sqlite search is doing `SELECT COUNT(DISTINCT path) FROM DerivationOutputs WHERE id="out"`, but that lists *everything* and I just prefer the current lookup format.

**Comment by Gobidev** [comment]
> The other option for an sqlite search is doing SELECT COUNT(DISTINCT path) FROM DerivationOutputs WHERE id="out", but that lists everything

How does that count compare to the other methods?

**Comment by coolGi2007** [comment]
Doing a quick test:

| Type | Speed | Packages |
| --- | --- | --- |
| Original | 2.9 ms ± 0.1 ms | N/A |
| `SELECT COUNT(path) FROM ValidPaths WHERE ultimate=1` (this pr) | 5.3 ms ± 0.9 ms | 832 |
| `SELECT COUNT(DISTINCT path) FROM DerivationOutputs WHERE id="out"` | 15.1 ms ± 3.0 ms | 8521 |

<br>

If I remove the `DISTINCT` to make the other lookup go faster gets quicker times, but the amount of packages also go way up.
| Type | Speed | Packages |
| --- | --- | --- |
| `SELECT COUNT(path) FROM DerivationOutputs WHERE id="out"` | 6.4 ms ± 0.3 ms | 16333 |

<br>

`nix-store -q --requisites /run/current-system/sw | wc -l`: 1306\
`nix-store -q --requisites ~/.nix-profile | wc -l`: 1171\
(note, the packages in current system and nix profile will have overlap)

Getting the overlap by doing the following:
```bash
nix-store -q --requisites /run/current-system/sw > /tmp/pkgs_system.txt
nix-store -q --requisites ~/.nix-profile > /tmp/pkgs_profile.txt
awk 'FNR==NR {a[$0]++; next} !($0 in a)' /tmp/pkgs_profile.txt /tmp/pkgs_system.txt
```
(on my system being 651)\
Then subtracting that from the sum gives 1826 on my system. (again, remember that this number includes much more than just packages)

**Comment by Gobidev** [comment]
Thanks a lot for the extensive comparison :)

Based on your numbers, your current method seems to be the one to go with. If there was a way to include dependencies in the count, that would be preferable, but that might not be possible with this approach.

I tested your method on a stock NixOS system, and it returns `313` installed packages, which is a believable number. I will test some more scenarios in which Nix might be used tomorrow.

Thanks again for the amount of work you put into this, it is nice to finally see Nix support being added to libmacchina^^

**Review by Gobidev — APPROVED** [review]

**Comment by grtcdr** [comment]
I'd include the dependencies to avoid the inevitable, unwanted reports we'd undoubtedly receive from users. I'm familiar with the basic concepts of Nix, but the intricacies of it escape me, so if the majority agrees that including just the base packages is a better index then I'm all for it.

**Comment by coolGi2007** [comment]
How about something like `SELECT COUNT(path) FROM ValidPaths WHERE sits IS NOT NULL`\
It seems to give the correct results with dependacies.

On a fresh install of nix on another distro, it outputs 0. Then once I install cargo (which has 63 dependacies, that it says at the top of the output when its downloading), it outputs 64 as expected.\
I think this is what we've been looking for!

Sorry, not sure how I didnt notice that before. I guess doing my tests on a non-nix system is easier to read as there isn't soo many packages installed.

**Comment by coolGi2007** [comment]
Yeah, this is wayy better!\
Doing the same thing to get the amount of packages on both the current generations (printing the output of the `nix-store` thingie and getting the amount of non-duplicates) gives 1826. Then with this new method it gives 1870.

This slightly larger number is understandable as the new sqlite lookup includes every generation (which won't add any on my system due to me cleaning old generations recently), as well as packages on your system but not on the current generation (ie from `nix-shell`)

**Comment by Gobidev** [comment]
I have tested this new method and found the following:

On an Arch installation with Nix installed in multi-user mode, I get `67` installed packages after installing `cargo` and `pfetch-rs` with nix, which looks good.

On an Arch installation with Nix installed in single-user mode, I get `28` installed packages after installing `wget` and `cargo`, but it stays at `28` after uninstalling `wget`.

On a NixOS installation, i get `331` installed packages (it was `314` with the previous method), while `nix-store -q --requisites /run/current-system/sw | wc -l` alone returns `929`.

This method is still not perfect but imo the best we've had so far.

**Comment by coolGi2007** [comment]
Remember to clear your cache (`nix-collect-garbage -d`)

**Comment by grtcdr** [comment]
I'm happy with what we've got - thank you so much for the quality work - shall we get this merged?

**Comment by coolGi2007** [comment]
Yeah, sounds good!

**Review by grtcdr — APPROVED** [review]

**Comment by coolGi69** [comment]
Fyi, the wiki should probs be updated to say that the Nix package manager is supported as well~
https://github.com/Macchina-CLI/macchina/wiki/Package-Manager-Support

**Comment by grtcdr** [comment]
@coolGi69 Good catch, done!

**Pull request merged**